### ISSUE TITLE
DynamoDB DeleteTable returning 500 response but only when we run in a GH action

### ISSUE THREAD

**Issue by adam-fowler** [localstack/localstack#2596]
# Type of request: This is a ...
[ x ] bug report
[ ] feature request

# Detailed description
I've been using Localstack in the CI for the community developed [AWS SDK Swift](https://github.com/swift-aws/aws-sdk-swift) for a while. In general when I've come across an issue I have been able to replicate the issue using the aws cli and then report as such. 

I have a number of tests which create a DynamoDB table, do something with table and then delete it. In the last few days the table deletion has started failing but only when running in a Github Action. You can see it [here](https://github.com/adam-fowler/dynamodb-codable/runs/792431390?check_suite_focus=true#step:5:3652). It has been returning a status code 500 with a body `{}`.

I cannot replicate this locally, I have compared the http request from aws-cli and AWS SDK Swift and they both look the same. The request to the REST interface is as follows

```
POST http://localstack:4566/
Headers: [
  ("Content-Type", "application/x-amz-json-1.0"), 
  ("User-Agent", "AWSSDKSwift/5.0"), 
  ("x-amz-target", "DynamoDB_20120810.DeleteTable"),
  ("X-Amz-Date", "20200621T110705Z"), 
  ("host", "localstack:4566"), 
  ("x-amz-content-sha256", "d1f2b34a00570e27cda823fe048ec5d0891210b37a80765766d3af74d60106bc"),
  ("Authorization", "AWS4-HMAC-SHA256 Credential=foo/20200621/us-east-1/dynamodb/aws4_request, SignedHeaders=content-type;host;user-agent;x-amz-content-sha256;x-amz-date;x-amz-target, Signature=5cfb62132afa2b2adbfe7f69f86d1c1741d97920cc1a0902ced18f5595fdb525")
]
Body: {"TableName":"awssdkswift-testcreatedelete"}
```

Have you any idea why this would work locally but not work while running in a GH action?

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)
I included the Swift code below

```
    func createTable(name: String) -> EventLoopFuture<Void> {
        let input = DynamoDB.CreateTableInput(
            attributeDefinitions: [.init(attributeName: "id", attributeType: .s)],
            keySchema: [.init(attributeName: "id", keyType: .hash)],
            provisionedThroughput: .init(readCapacityUnits: 5, writeCapacityUnits: 5),
            tableName: name
        )
        return Self.dynamoDB.createTable(input)
            .map { response in
                XCTAssertEqual(response.tableDescription?.tableName, name)
                return
        }
    }
    
    func deleteTable(name: String) -> EventLoopFuture<Void> {
        let input = DynamoDB.DeleteTableInput(tableName: name)
        return Self.dynamoDB.deleteTable(input).map { _ in }
    }
    
    func testCreateDelete() {
        let tableName = "awssdkswift-testcreatedelete"
        let response = createTable(name: tableName)
            .flatMap { _ in
                self.deleteTable(name: tableName)
        }
        XCTAssertNoThrow(try response.wait())
    }
```

┆Issue is synchronized with this [Jira Bug](https://localstack.atlassian.net/browse/LOC-201) by [Unito](https://www.unito.io/learn-more)

**Comment by whummer** [comment]
Thanks for reporting @adam-fowler . My assumption would be that this could be an issue related to resource limitations in GH actions..? Please note that for the DynamoDB API we're using _DynamoDBLocal_ under the covers, which is a Java program running in a JRE process. Perhaps there are issues with spawning the JRE, or potentially the process is running out of memory..?

Can you please share your full configuration - have you tried customizing the `SERVICES` environment variable, to start only the services that are required for your test? Thanks

**Comment by adam-fowler** [comment]
I am already running with SERVICES set to dynamodb. Here is my GH action yml.

```
name: CI

on:
  push:
    branches:
    - master
  pull_request:
    branches:
    - master

jobs:
  linux:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        image: ['swift:5.1', 'swift:5.2']
    services:
      localstack:
        image: localstack/localstack
        env:
          SERVICES: 'dynamodb'
    container:
      image: ${{ matrix.image }}
      env:
        DYNAMODB_ENDPOINT   : "http://localstack:4566"
        AWS_ENABLE_LOGGING  : "true"
    steps:
    - name: Checkout
      uses: actions/checkout@v1
      with:
        fetch-depth: 1
    - name: Install dependencies
      run: |
        apt-get update -qq
        apt-get install -q -y tzdata zlib1g-dev
    - name: Test
      run: swift test --enable-test-discovery --sanitize=thread
```
Locally on my mac, where my code is working, I am running localstack as follows
```
docker run -d -p 4566-4597:4566-4597 -p 8080:8080 -e SERVICES='dynamodb' localstack/localstack
```

**Comment by adam-fowler** [comment]
I've managed to replicate this with awscli in a github action with the following
```
name: AWSCLI

on: [push]

jobs:
  linux:
    runs-on: ubuntu-latest
    services:
      localstack:
        image: localstack/localstack
        env:
          SERVICES: 'dynamodb'
        ports:
          - "4566:4566"
          - "8080:8080"
    env:
      AWS_ACCESS_KEY_ID   : "foo"
      AWS_SECRET_ACCESS_KEY: "bar"
      AWS_DEFAULT_REGION  : "eu-west-1"
    steps:
    - name: Install dependencies
      run: |
        sudo apt-get update -qq
        sudo apt-get install -q -y curl unzip python3-setuptools python3-pip
        python3 -m pip install awscli
        aws --version
    - name: Test
      run: |
        curl localhost:8080/health
        aws --endpoint http://localhost:4566 --debug dynamodb create-table --table-name awssdkswift-testtable --attribute-definitions AttributeName=id,AttributeType=N --key-schema AttributeName=id,KeyType=HASH --provisioned-throughput ReadCapacityUnits=1,WriteCapacityUnits=1
        aws --endpoint http://localhost:4566 --debug dynamodb delete-table --table-name awssdkswift-testtable
```

**Comment by rb16** [comment]
same issue here 


Docker compose - 
```
services:
  localstack:
    networks:
      - shared
    image: localstack/localstack
    container_name: localstack
    ports:
      - "4566-4599:4566-4599" # ports of the different services
      - "9000:9000"
    environment:
      - SERVICES=dynamodb
      - DEFAULT_REGION=ap-southeast-1
      - HOSTNAME_EXTERNAL=localstack
      - HOSTNAME=localstack
      - PORT_WEB_UI=9000
      - DOCKER_HOST=unix:///var/run/docker.sock
      - START_WEB=1
      - DATA_DIR=/tmp/localstack/data

    volumes:
      - './.localstack:/tmp/localstack'
      - '/var/run/docker.sock:/var/run/docker.sock'
```


Delete command:- 
```
 aws  --endpoint-url=http://localhost:4566 dynamodb delete-table --table-name my_Table

An error occurred (500) when calling the DeleteTable operation (reached max retries: 9): 
```


Container logs:- 
```
localstack        | 2020-06-21T19:09:31:WARNING:localstack.utils.server.http2_server: Error in proxy handler for request POST http://localhost:4566/: Parameter validation failed:
localstack        | Invalid type for parameter EventSourceArn, value: None, type: <class 'NoneType'>, valid types: <class 'str'> Traceback (most recent call last):
localstack        |   File "/opt/code/localstack/localstack/utils/server/http2_server.py", line 53, in index
localstack        |     result = await run_sync(handler, request, data)
localstack        |   File "/opt/code/localstack/localstack/utils/async_utils.py", line 46, in run_sync
localstack        |     return await loop.run_in_executor(THREAD_POOL, copy_context().run, func, *args)
localstack        |   File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
localstack        |     result = self.fn(*self.args, **self.kwargs)
localstack        |   File "/opt/code/localstack/localstack/services/generic_proxy.py", line 550, in handler
localstack        |     response = modify_and_forward(method=method, path=path_with_params, data_bytes=data, headers=headers,
localstack        |   File "/opt/code/localstack/localstack/services/generic_proxy.py", line 393, in modify_and_forward
localstack        |     updated_response = update_listener.return_response(**kwargs)
localstack        |   File "/opt/code/localstack/localstack/services/dynamodb/dynamodb_listener.py", line 285, in return_response
localstack        |     self.delete_all_event_source_mappings(table_arn)
localstack        |   File "/opt/code/localstack/localstack/services/dynamodb/dynamodb_listener.py", line 407, in delete_all_event_source_mappings
localstack        |     result = lambda_client.list_event_source_mappings(EventSourceArn=table_arn)
localstack        |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/botocore/client.py", line 316, in _api_call
localstack        |     return self._make_api_call(operation_name, kwargs)
localstack        |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/botocore/client.py", line 607, in _make_api_call
localstack        |     request_dict = self._convert_to_request_dict(
localstack        |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/botocore/client.py", line 655, in _convert_to_request_dict
localstack        |     request_dict = self._serializer.serialize_to_request(
localstack        |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/botocore/validate.py", line 297, in serialize_to_request
localstack        |     raise ParamValidationError(report=report.generate_report())
localstack        | botocore.exceptions.ParamValidationError: Parameter validation failed:
localstack        | Invalid type for parameter EventSourceArn, value: None, type: <class 'NoneType'>, valid types: <class 'str'>
localstack        | 

```

**Comment by usmangani1** [comment]
@rb16  Can you pull the latest docker image and try once again ? Please let us know in case if you see the same issue .

**Comment by adam-fowler** [comment]
@usmangani1 I re-ran my github action and it still fails, although in a slightly different way. Initially awscli receives a 500 error but on retrying it gets the following error. 
```
ResourceNotFoundException: An error occurred (ResourceNotFoundException) when calling the DeleteTable operation: Cannot do operations on a non-existent table
```
You can see the full log here https://github.com/adam-fowler/localstack-awscli/runs/796946379

**Comment by lsena** [comment]
Having the same error using aws CodeBuild.

`ResourceNotFoundException: An error occurred (ResourceNotFoundException) when calling the DeleteTable operation: Cannot do operations on a non-existent table`

**Comment by macnaughton** [comment]
Had a similar error with deleting tables; downgrading to 0.11.2 has resolved it for now.

**Comment by whummer** [comment]
Thanks for the update @lsena @macnaughton @adam-fowler . We recently pushed a fix that upgrades the DynamoDB backend to the latest version. Can you please give it another try with the `latest` version of the Docker image? Please report here if the problem persists. Thanks

**Comment by adam-fowler** [comment]
@whummer still getting the `ResourceNotFoundException` error

**Comment by robertbullen** [comment]
I ran into the most recent manifestation of this bug ('ResourceNotFoundException: Cannot do operations on a non-existent table') as well. In my case I'm running unit tests from Node.js with the JavaScript AWS SDK, connecting to localstack/localstack-light running within Docker Desktop, all on my MacBook Pro.

So this bug isn't necessarily specific to GitHub actions. But my scenario does have something in common with @adam-fowler's: my unit tests create a DynamoDB table, exercise said table, and then delete it. Perhaps the short, thrashy lifetime of the table has something to do with it?

**Comment by whummer** [comment]
Thanks for the update @robertbullen @adam-fowler @macnaughton @lsena . We have recently pushed a fix in #3717 which should fix the Docker startup in Github actions (which might be related to the issues you've been facing). Can you please pull the `latest` Docker image and give it anothery try? Please let us know if the problem still persists. Thanks!
